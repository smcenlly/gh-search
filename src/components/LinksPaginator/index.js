export { default as LinksPaginator } from "./LinksPaginator";
