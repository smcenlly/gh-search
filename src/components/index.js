export * from "./UserSearch";
export * from "./Repos";
export * from "./LinksPaginator";
