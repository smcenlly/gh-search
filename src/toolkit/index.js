export * from "./Loader";
export * from "./Message";
export * from "./Button";
