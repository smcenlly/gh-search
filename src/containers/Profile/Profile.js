import _ from "lodash";
import i18n from "i18next";
import React, { Component } from "react";
import styled from "styled-components";
import { Repos, UserSearch, LinksPaginator } from "../../components";
import { Loader, Message } from "../../toolkit";
import ReposService from "../../services/repos";

const Wrapper = styled.div`
    display: flex;
    flex-flow: column;
    height: 100%;
`;

const Header = styled.h1`
    color: #fff;
    line-height: 1.5;
    margin: 0;
    background: linear-gradient(90deg, rgb(65, 106, 162), rgb(125, 96, 196));
    padding: 50px 20px 10px;
`;

class Profile extends Component {
    constructor(props) {
        super(props);

        this.state = {
            searchInputValue: "",
            lastCommitedValue: "",
            isSubmitting: false,
            error: null,
            page: null
        };

        this.fetchRepos = _.debounce(this.fetchRepos, 250);
    }

    fetchRepos = async (userName, pageNumber) => {
        this.setState({ page: null, error: null, isSubmitting: true });

        try {
            const page = await ReposService.getPage(userName, pageNumber);

            this.setState({
                page,
                searchInputValue: "",
                lastCommitedValue: userName
            });
        } catch (err) {
            this.setState({ error: err });
        }

        this.setState({ isSubmitting: false });
    };

    handleChange = ({ target: { value } }) => {
        this.setState({ searchInputValue: value });
    };

    handleSubmit = () => {
        return this.fetchRepos(this.state.searchInputValue);
    };

    render() {
        if (process.env.NODE_ENV === "development" && this.state.error) {
            console.error(this.state.error);
        }

        return (
            <Wrapper>
                <Header>{i18n.t("profile.title")}</Header>
                <UserSearch
                    value={this.state.searchInputValue}
                    onChange={this.handleChange}
                    isSubmitting={this.state.isSubmitting}
                    onSubmit={this.handleSubmit}
                />
                {this.state.page ? (
                    <LinksPaginator
                        links={this.state.page && this.state.page.links}
                        onPaginate={(page) =>
                            this.fetchRepos(this.state.lastCommitedValue, page)
                        }
                    />
                ) : null}

                {this.state.error ? (
                    <Message level={Message.LEVEL.ERROR}>
                        {i18n.t("userSearch.error")}
                    </Message>
                ) : null}

                {this.state.isSubmitting ? (
                    <Loader align={Loader.ALIGN.LEFT} />
                ) : null}

                {this.state.page && this.state.page.data ? (
                    <Repos repos={this.state.page.data} />
                ) : null}
            </Wrapper>
        );
    }
}

export default Profile;
