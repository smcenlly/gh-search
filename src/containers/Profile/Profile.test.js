import React from "react";
import { shallow } from "enzyme";
import { Loader, Message } from "../../toolkit";
import { Repos, UserSearch } from "../../components";
import Profile from "./Profile";
import ReposService from "../../services/repos";

describe("Profile", () => {
    let wrapper,
        getUserSearch = () => wrapper.find(UserSearch).get(0),
        getLoader = () => wrapper.find(Loader).get(0),
        getMessage = () => wrapper.find(Message).get(0),
        commitUserSearch = () => {
            getUserSearch().props.onChange({ target: { value: "Sample" } });
            return getUserSearch().props.onSubmit();
        };

    beforeEach(() => {
        wrapper = shallow(<Profile />);
        jest.useFakeTimers();
    });

    afterEach(() => {
        jest.clearAllMocks();
        jest.clearAllTimers();
    });

    it("renders correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe("Given submitting UserSearch", () => {
        it("should show <Loader />", () => {
            jest.spyOn(ReposService, "getAll").mockImplementationOnce(
                () => new Promise((resolve) => setTimeout(resolve))
            );

            commitUserSearch();

            expect(wrapper).toContainMatchingElements(1, Loader);
        });

        describe("when successful", () => {
            it("shows list of repos", async () => {
                const repos = [{ id: 1 }, { id: 2 }];
                jest.spyOn(ReposService, "getAll").mockImplementationOnce(() =>
                    Promise.resolve(repos)
                );

                await commitUserSearch();

                expect(wrapper).toContainMatchingElements(1, Repos);
                expect(wrapper.find(Repos).first()).toHaveProp({
                    repos
                });
            });
        });
        describe("when unsuccessful", () => {
            it("shows error <Message />", async () => {
                jest.spyOn(ReposService, "getAll").mockImplementationOnce(() =>
                    Promise.reject("error")
                );

                await commitUserSearch();

                expect(wrapper).toContainMatchingElements(1, Message);
                expect(wrapper.find(Message).first()).toHaveProp({
                    level: Message.LEVEL.ERROR
                });
            });
        });
    });
});
