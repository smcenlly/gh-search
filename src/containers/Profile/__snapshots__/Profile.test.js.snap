// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders correctly 1`] = `
<styled.div>
  <styled.h1 />
  <UserSearch
    isSubmitting={false}
    onChange={[Function]}
    onSubmit={[Function]}
    value=""
  />
</styled.div>
`;
